libcache/cache.c

```diff
@@ -0,0 +1,391 @@
+	}
+
+	for (; i < cache->numSets; ++i) {
+		set = cache_setInit();
+
+		if (set == NULL) {
+			for (; j < i; ++j) {
+				free(cache->sets[j].tags);
+			}
+			free(cache->sets);
+			free(cache);
+
+			return NULL;
+		}
+		cache->sets[i] = *set;
+
+		cache->sets[i].count = 0;
+
+		free(set);
+	}
+
+	cache->offBitNum = log2(lineSize);
+	off_t setBits = log2(cache->numSets);
+	off_t tagBits = LIBCACHE_ADDR_WIDTH - setBits - cache->offBitNum;
+
+	cache->tagMask = cache_genMask(tagBits);
+	cache->setMask = cache_genMask(setBits);
+	cache->offMask = cache_genMask(cache->offBitNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i = 0;
+
+	for (; i < cache->numSets; ++i) {
+		if (cache->sets != NULL) {
+			free(cache->sets[i].tags);
+		}
+	}
+
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const off_t offset, int policy)
+{
+	int i = 0;
+	ssize_t wrt = 0;
+	cacheline_t *temp = NULL, *oldest = NULL, **linePtr = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		for (; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->lines[i].isValid != '1') {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		switch (policy) {
+			case 0: {
+				/* write-back */
+				set->lines[i].isDirty = '1';
+				break;
+			}
+			case 1: {
+				/* write-through */
+				wrt = (*writeCb)(offset, temp->data, sizeof *(temp->data), 1);
+				break;
+			}
+			default: {
+				break;
+			}
+		}
+
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->tags[i] == NULL) {
+				set->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+		set->count++;
+	}
+	else {
+		oldest = set->timestamps;
+		LIST_REMOVE(&set->timestamps, oldest);
+
+		/* write-back */
+		if (oldest->isDirty == '1') {
+			wrt = (*writeCb)(offset, oldest->data, sizeof *(oldest->data), 1);
+		}
+
+		for (; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->lines[i].tag == oldest->tag) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		switch (policy) {
+			case 0: {
+				/* write-back */
+				set->lines[i].isDirty = '1';
+				break;
+			}
+			case 1: {
+				/* write-through */
+				wrt = (*writeCb)(offset, oldest->data, sizeof *(oldest->data), 1);
+				break;
+			}
+			default: {
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, oldest);
```

Comment:
Code duplication with the case above - refactor this to move this after if-else

Reply:
Thank you, done